x/ccv/provider/keeper/proposal.go

```diff
@@ -179,3 +259,72 @@ func (k Keeper) IteratePendingClientInfo(ctx sdk.Context) {
 		}
 	}
 }
+
+// SetPendingStopProposal sets the consumer chain ID for the given timestamp
+func (k Keeper) SetPendingStopProposal(ctx sdk.Context, chainID string, timestamp time.Time) {
+	store := ctx.KVStore(k.storeKey)
+	store.Set(types.PendingStopProposalKey(timestamp, chainID), []byte{})
+}
+
+// GetPendingStopProposal returns a boolean if a pending stop proposal exists for the given consumer chain ID and the timestamp
+func (k Keeper) GetPendingStopProposal(ctx sdk.Context, chainID string, timestamp time.Time) bool {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(types.PendingStopProposalKey(timestamp, chainID))
+
+	return bz != nil
+}
+
+// DeletePendingStopProposal deletes the stop proposal for the given consumer chain ID and given timestamp
+func (k Keeper) DeletePendingStopProposal(ctx sdk.Context, chainID string, timestamp time.Time) {
+	store := ctx.KVStore(k.storeKey)
+	store.Delete(types.PendingStopProposalKey(timestamp, chainID))
+}
+
+// IteratePendingStopProposal iterates over the pending stop proposals in order and stop the chain if the stop time has passed,
+// otherwise it will break out of loop and return.
+func (k Keeper) IteratePendingStopProposal(ctx sdk.Context) {
+	store := ctx.KVStore(k.storeKey)
+	iterator := sdk.KVStorePrefixIterator(store, []byte(types.PendingStopProposalKeyPrefix+"/"))
+	defer iterator.Close()
+
+	if !iterator.Valid() {
+		return
+	}
+
+	for ; iterator.Valid(); iterator.Next() {
+		suffixKey := iterator.Key()
+		// splitKey contains the bigendian time in the first element and the chainID in the second element
+		splitKey := strings.Split(string(suffixKey), "/")
+
+		timeNano := binary.BigEndian.Uint64([]byte(splitKey[0]))
+		stopTime := time.Unix(0, int64(timeNano))
+		chainID := string([]byte(splitKey[1]))
+
+		if ctx.BlockTime().After(stopTime) {
+			k.StopConsumerChain(ctx, chainID, false)
+			k.DeletePendingStopProposal(ctx, chainID, stopTime)
```

Comment:
Shouldn't we delete also the `PendingClientInfo`s when iterating through them (see `IteratePendingClientInfo`)?

Reply:
Agree. Should we create a separate PR for the changes touching the `PendingClientInfos` states?

Reply:
I think in general we should not be deleting while iterating